### PR TITLE
publish-commit-bottles: optionally push to pull request branch.

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
Rather than pulling bottles and pushing to master, pull bottles, push the bottle commit to the PR branch for automatic merging by GitHub and add a label to avoid rerunning `brew test-bot` on the various formulae again.

This also allows us to use the GitHub merge queue so run on `merge_group` events too.

CC @carlocab as we'd talked about this

### CODE REVIEW

**Review by carlocab — DISMISSED** [review]
One thing we might want to consider is to teach `test-bot` to ignore formulae which already have a separate bottle commit on the PR branch, but build ones for those still missing them. This might get us to something like incremental rebuilds on reruns.

File: .github/workflows/publish-commit-bottles.yml
Comment (ZhongRuoyu) [inline .github/workflows/publish-commit-bottles.yml:97]:
```suggestion
        run: gh pr edit --add-label CI-published-commit-bottles ${{github.event.inputs.pull_request}}
```

https://cli.github.com/manual/gh_pr_edit

**Comment by ZhongRuoyu** [comment]
Sounds worth trying out! Question: wouldn't the bottle commit dismiss existing approvals?

**Review by ZhongRuoyu — COMMENTED** [review]
Worth trying out!

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:81]: @carlocab This is unneeded if you're already passing `--clean`:

https://github.com/Homebrew/brew/blob/8a5e1697a64ab8b694fc1d4fd6c8e948fe6950be/Library/Homebrew/dev-cmd/pr-pull.rb#L443-L445

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:72]: @carlocab I'll check before merging.

**Comment by MikeMcQuaid** [comment]
> Question: wouldn't the bottle commit dismiss existing approvals?

@ZhongRuoyu If so: I'll adjust the branch protection rules to avoid this 👍🏻

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:97]:
```suggestion
        run: gh pr edit --add-label CI-published-bottle-commits ${{github.event.inputs.pull_request}}
```
Maybe? Fine either way tho.

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:72]: Hmm, added `gh` to `ghcr.io/homebrew/ubuntu22.04` but it's not finding it. Any ideas here anyone?

File: .github/workflows/tests.yml
Comment (carlocab) [inline .github/workflows/tests.yml:91]:
```suggestion
            if (label_names.includes('CI-published-bottle-commits')) {
              console.log('CI-published-bottle-commits label found. Skipping tests job.')
```
Since you took my rename suggestion above.

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:72]: @ZhongRuoyu good catch. I don't think I realised we weren't using `master` images here. Perhaps we should?

Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:72]: We should be using `master` images:

https://github.com/Homebrew/homebrew-core/blob/9fbda613cf9848e0f84f0a84acf429201ab57867/.github/workflows/tests.yml#L26-L27

Comment (ZhongRuoyu) [reply .github/workflows/publish-commit-bottles.yml:72]: Ah, sorry, I was looking at the wrong one. Should be  https://github.com/Homebrew/brew/pull/14982 then.

Comment (ZhongRuoyu) [reply .github/workflows/publish-commit-bottles.yml:72]: For `master` branch, here, in another workflow: https://github.com/Homebrew/brew/blob/master/.github/workflows/tests.yml#L180-L190

```yaml
      - name: Deploy the Docker image to GitHub Packages and Docker Hub
        if: github.ref == 'refs/heads/master'
        run: |
          echo ${{secrets.HOMEBREW_BREW_GITHUB_PACKAGES_TOKEN}} |
            docker login ghcr.io -u BrewTestBot --password-stdin
          docker tag brew "ghcr.io/homebrew/ubuntu22.04:master"
          docker push "ghcr.io/homebrew/ubuntu22.04:master"
          echo ${{secrets.HOMEBREW_BREW_DOCKER_TOKEN}} |
            docker login -u brewtestbot --password-stdin
          docker tag brew "homebrew/ubuntu22.04:master"
          docker push "homebrew/ubuntu22.04:master"
```

Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:72]: Kinda weird for that to be in `tests.yml`, but ok. In any case, [`gh` does still need `gh auth login`](https://github.com/Homebrew/homebrew-core/actions/runs/4428620069/jobs/7770072493#step:4:12).
```
Welcome to GitHub CLI!
To authenticate, please run `gh auth login`.
```

**Review by ZhongRuoyu — COMMENTED** [review]
These too. (⌘F says that's all.)

**Comment by ZhongRuoyu** [comment]
Two more things (for now):

1. Maybe, we want to consider changing the wording of comments posted [here](https://github.com/Homebrew/homebrew-core/blob/publish_commit_bottles_to_pr/.github/workflows/publish-commit-bottles.yml#L33-L40)? Currently they read either ":shipit: @BrewTestBot has triggered a merge" or ":robot: A scheduled task has triggered a merge".

2. Will this ([L83-L91](https://github.com/Homebrew/homebrew-core/blob/publish_commit_bottles_to_pr/.github/workflows/publish-commit-bottles.yml#L83-L91); couldn't comment there) still push to `master`...
```yaml
      - name: Push commits
        uses: Homebrew/actions/git-try-push@master
        with:
          token: ${{secrets.HOMEBREW_GITHUB_PUBLIC_REPO_TOKEN}}
          directory: ${{steps.set-up-homebrew.outputs.repository-path}}
        env:
          GIT_COMMITTER_NAME: BrewTestBot
          GIT_COMMITTER_EMAIL: 1589480+BrewTestBot@users.noreply.github.com
          HOMEBREW_GPG_PASSPHRASE: ${{ secrets.BREWTESTBOT_GPG_SIGNING_SUBKEY_PASSPHRASE }}
```

... because of this? https://github.com/Homebrew/actions/blob/master/git-try-push/action.yml#L18-L21

```yaml
  branch:
    description: Git branch
    required: false
    default: master
```

Even if not, perhaps we'd better be more explicit.

**Comment by MikeMcQuaid** [comment]
> We could enable automerge before pushing the bottle commit, but that could also happen before automerge is enabled, but after this workflow starts

That's a good idea 👍🏻

File: .github/workflows/tests.yml
Comment (carlocab) [inline .github/workflows/tests.yml:43]: Looks like this worked, but you need to `cd` to homebrew/core.

**Comment by carlocab** [comment]
> > We could enable automerge before pushing the bottle commit, but that could also happen before automerge is enabled, but after this workflow starts
> 
> 
> 
> That's a good idea 👍🏻 

Wouldn't this result in a PR getting merged before the bottle commit is pushed if it was previously approved? Or is that fixed by `merge_group`?

**Comment by MikeMcQuaid** [comment]
> Wouldn't this result in a PR getting merged before the bottle commit is pushed if it was previously approved?

Yes, that's true. Need to enable auto-merge immediately after pushing I guess!

**Review by carlocab — COMMENTED** [review]
Happy with where this is going, but I think it would be useful to gate these changes behind a PR label so that we can test it out on a few PRs before rolling it out fully.

Then, once we start merging most PRs this way, we can also use a label for the old merge flow for PRs that don't have `Allow edits from maintainers` enabled (e.g. org forks).

**Review by Bo98 — COMMENTED** [review]
Been thinking about it a bit and wondering how we could use merge queues to handle this. Turns out merge groups are a bit more limited than I thought.

Will take a proper look at this at the weekend but here's a couple questions from a quick think through.

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:34]: @Bo98 In this case as-is: the default. We should start using merge commits in this repository (and homebrew-cask) along with sharing `Formula`/`Casks` to make the Git operations on these repositories more performance so by the time I merge this I expect: merge commits.

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:82]: @Bo98 Alternatively, we can (and should) care less about commit message/format pedantry, particularly if/when we're moving to merge commits where it matters much less about narrowing things down to a single commit.

**Comment by MikeMcQuaid** [comment]
> Been thinking about it a bit and wondering how we could use merge queues to handle this. Turns out merge groups are a bit more limited than I thought.

@bo98 Yes, this gets us on the way to be able to use merge queues for some flows but you cannot push or modify commits as part of the merge queue. We should not expect that to change.

Expanding on the above: the homebrew/core CI flow is (and I say this as the person who is primarily to blame for this 😅) very unusual, makes it hard to rely on GitHub's security and workflow features (shout out to all our PRs with bottles being "closed" rather than "merged", something GitHub is never likely to let us fix), makes the repository network size balloon and stops us from being able to repeatedly rerun builds and reuse existing work.

To be very clear: I don't expect to just merge this PR as is in the next few days and I expect there will be follow-up changes required to fix and improve these workflows.

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [reply .github/workflows/publish-commit-bottles.yml:82]: That would be disappointing - `git blame` is actually really useful here, compared to sometimes being horrid to use on Homebrew/brew. But I'll consider the rest of the concept first.

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [reply .github/workflows/publish-commit-bottles.yml:34]: It's not clear what "the default" meant considering merge commits isn't currently enabled and the default on the web interface is just the last one you used.

**Comment by carlocab** [comment]
> I feel like you are a working record because this sounds like the first time I heard this, sorry 😂

It wasn't that long ago! 😅 https://github.com/Homebrew/homebrew-core/pull/125556#pullrequestreview-1342839829

> Definitely game to land this and #125595 sooner rather than later for experimentation.

I think that this is very close to a state where I think it's worth landing to experiment with.

> How do you see the PR flag working? A label? Something different?

Yes, by "flag" I actually meant "label" (I got it right the first time I said it though).

> Feel free to commit directly to this PR (and #125595) if that's easier than leaving comments here.

Sure, will work on this PR tomorrow.

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:34]: I _think_ `gh` requires specification of a merge strategy to use `gh pr merge`. i.e. we must pass one of `--squash`, `--rebase`, or `--merge`.

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:82]: > I'm also not entirely sure how `brew extract`/`FormulaVersions` behaves when a version bump is split into several commits but that can be adjusted if need be.

Yea, I think we may break `brew extract` and `FormulaVersions`. Though I think the former will be broken because it relies on the latter. Breaking `FormulaVersions` would not be good -- we use it for some audits.

**Comment by carlocab** [comment]
Force-pushed to resolve merge conflict.

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:23]: It turned out to be a bit fiddly to check the PR for labels, so let's use a workflow dispatch input instead. We can teach `brew pr-publish` to pass the right argument here if the PR has a given label if desired.

**Comment by carlocab** [comment]
Odd:
```
Message: Resource not accessible by integration, Locations: [{Line:1 Column:54}]
```

**Comment by carlocab** [comment]
Token needs different permissions:
```
GraphQL: Your token has not been granted the required scopes to execute this query. The 'login' field requires one of the following scopes: ['read:org'], but your token has only been granted the: ['public_repo'] scopes. Please modify your token's scopes at: https://github.com/settings/tokens., Your token has not been granted the required scopes to execute this query. The 'name' field requires one of the following scopes: ['read:org', 'read:discussion'], but your token has only been granted the: ['public_repo'] scopes. Please modify your token's scopes at: https://github.com/settings/tokens., Your token has not been granted the required scopes to execute this query. The 'slug' field requires one of the following scopes: ['read:org', 'read:discussion'], but your token has only been granted the: ['public_repo'] scopes. Please modify your token's scopes at: https://github.com/settings/tokens.
```

**Review by carlocab — DISMISSED** [review]
Modulo token issues and cleanup, the only change that this will really have that isn't opt-in is `merge_group`. Happy to land this for experimentation once token issues and cleanup is done.

**Comment by MikeMcQuaid** [comment]
> It wasn't that long ago! 😅 [#125556 (review)](https://github.com/Homebrew/homebrew-core/pull/125556#pullrequestreview-1342839829)

Whoops, sorry, missed that!


> Modulo token issues and cleanup, the only change that this will really have that isn't opt-in is `merge_group`. Happy to land this for experimentation once token issues and cleanup is done.

Awesome, thanks @carlocab, yeh, please feel free to land when you're happy with it. The `merge_group` change is essentially a no-op until we enable the merge queue branch protection so should be safe as-is.

**Comment by carlocab** [comment]
I can clean this up, but I was hoping you could handle the token problem since I have no idea what to do with it 😅 In particular, I don't know which tokens have which scopes, etc.

**Comment by MikeMcQuaid** [comment]
> I can clean this up, but I was hoping you could handle the token problem since I have no idea what to do with it 😅 In particular, I don't know which tokens have which scopes, etc.

@carlocab sure thing will take a look!

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:80]: Yep, I've tried `GITHUB_TOKEN`, but it seems to not have the right permissions either.

Comment (MikeMcQuaid) [reply .github/workflows/publish-commit-bottles.yml:82]: @carlocab Working now. Needed to add `permissions`.

**Review by carlocab — APPROVED** [review]

**Comment by carlocab** [comment]
Needs `@Homebrew/tsc` to merge.

**Review by Rylan12 — APPROVED** [review]
This looks awesome, I'm very excited to see this in action!

If someone wants to get this merged tonight: I'll be around for several hours in case a TSC approval is needed for a revert. I'll try to periodically check but if it's urgent ping me on Slack so it sends me a notification.

**Pull request merged**
